cmd/snap-confine/snap-confine-args.c

```diff
@@ -0,0 +1,206 @@
+		}
+	}
+	// Parse arguments that we've got
+	int optind;
+	for (optind = 1; optind < argc; ++optind) {
+		// Look at all the options switches that start with the minus sign ('-')
+		if (argv[optind][0] == '-') {
+			// Handle option switches
+			if (strcmp(argv[optind], "--version") == 0) {
+				args->is_version_query = true;
+				// NOTE: --version short-circuits the parser to finish
+				break;
+			} else if (strcmp(argv[optind], "--classic") == 0) {
+				args->is_classic_confinement = true;
+			} else {
+				// Report unhandled option switches
+				err =
+				    sc_error_init(SC_ARGS_DOMAIN,
+						  SC_ARGS_ERR_USAGE,
+						  "unrecognized command line option: %s",
+						  argv[optind]);
+				goto out;
+			}
+		} else {
+			// Handle positional arguments
+			if (args->security_tag == NULL) {
+				// The first positional argument (that is not an option switch)
+				// becomes the security tag.
+				if (ignore_first_tag) {
+					// Unless we are called as ubuntu-core-launcher, then we
+					// just swallow and ignore that security tag altogether.
+					ignore_first_tag = false;
+					continue;
+				}
+				args->security_tag = strdup(argv[optind]);
+				if (args->security_tag == NULL) {
+					die("cannot allocate memory for security tag");
+				}
+			} else if (args->executable == NULL) {
+				// The second positional argument becomes the executable name.
+				args->executable = strdup(argv[optind]);
+				if (args->executable == NULL) {
+					die("cannot allocate memory for executable name");
+				}
+				// No more positional arguments are required.
+				// Stop the parsing process.
+				break;
+			}
+		}
+	}
+
+	if (args->is_version_query == false) {
+		// Ensure that we have the security tag
+		if (args->security_tag == NULL) {
+			err = sc_error_init(SC_ARGS_DOMAIN, SC_ARGS_ERR_USAGE,
+					    "application or hook security tag was not provided");
+			goto out;
+		}
+		// Ensure that we have the executable name
+		if (args->executable == NULL) {
+			err = sc_error_init(SC_ARGS_DOMAIN, SC_ARGS_ERR_USAGE,
+					    "executable name was not provided");
+			goto out;
+		}
+	}
+	// "shift" the argument vector left, except for argv[0], to "consume" the
+	// arguments that were scanned / parsed correctly.
+	int i;
+	for (i = 1; optind + i < argc; ++i) {
+		argv[i] = argv[optind + i];
+	}
+	argv[i] = NULL;
+
+	// Write the updated argc and argv back.
+	*argvp = argv;
+	*argcp = argc - optind;
+
+ out:
+	// Don't return anything in case of an error.
+	if (err != NULL) {
+		sc_cleanup_args(&args);
+	}
+	// Forward the error and return
+	sc_error_forward(errorp, err);
+	return args;
+}
+
+void sc_args_free(struct sc_args *args)
+{
+	if (args != NULL) {
+		free(args->security_tag);
+		args->security_tag = NULL;
+		free(args->executable);
+		args->executable = NULL;
+		free(args);
+	}
+}
+
+void sc_cleanup_args(struct sc_args **ptr)
+{
+	sc_args_free(*ptr);
+	*ptr = NULL;
+}
+
+bool sc_args_is_version_query(struct sc_args *args)
```

Comment:
It's not clear why we need all those helpers. If we cannot allocate arguments, we should return an error right then, instead of going through with an unallocated struct and checking that things are not completely broken on every call site.